explorer/interpreter/type_checker.cpp

```diff
@@ -4735,6 +4877,11 @@ auto TypeChecker::DeclareAliasDeclaration(Nonnull<AliasDeclaration*> alias,
 auto TypeChecker::TypeCheck(AST& ast) -> ErrorOr<Success> {
   ImplScope impl_scope;
   ScopeInfo top_level_scope_info = ScopeInfo::ForNonClassScope(&impl_scope);
+
+  // Track that `impl_scope` is the top-level `ImplScope`.
+  llvm::SaveAndRestore<decltype(top_level_impl_scope_)>
```

Comment:
Is there a need for decltype here instead of putting the actual type?

Reply:
I think `decltype` is clearer that we're storing the type of that member, and including the type would not be helpful to a reader of the code. (It's a pity that `llvm::SaveAndRestore` doesn't support CTAD; we should really be deducing the type here rather than specifying it.)

Reply:
FWIW, I think we tend not to use decltype this way. It seems unusual to use it here, when I don't think the type of top_level_impl_scope_ is templated (or is it? maybe I'm misunderstanding something? because to me, "decltype" is "the type of this will change")